File: app/src/main/java/com/faendir/lightning_launcher/multitool/badge/BadgeSetup.java
package com.faendir.lightning_launcher.multitool.badge;

import android.app.Activity;
import android.content.Intent;
import android.support.annotation.Keep;
import com.faendir.lightning_launcher.multitool.proxy.ActivityScreen;
import com.faendir.lightning_launcher.multitool.proxy.Container;
import com.faendir.lightning_launcher.multitool.proxy.EventHandler;
import com.faendir.lightning_launcher.multitool.proxy.JavaScript;
import com.faendir.lightning_launcher.multitool.proxy.PropertySet;
import com.faendir.lightning_launcher.multitool.proxy.ProxyFactory;
import com.faendir.lightning_launcher.multitool.proxy.Script;
import com.faendir.lightning_launcher.multitool.proxy.Shortcut;
import com.faendir.lightning_launcher.multitool.proxy.Utils;

/**
 * @author lukas
 * @since 08.07.18
 */
@Keep
public class BadgeSetup implements JavaScript.Setup, JavaScript.ActivityResult {
    static final String TAG_PACKAGE = "package";
    private final Utils utils;

    public BadgeSetup(Utils utils) {
        this.utils = utils;
    }

    @Override
    public void setup() {
        Script create = utils.installActivityResultScript();
        ActivityScreen screen = ProxyFactory.cast(utils.getActiveScreen(), ActivityScreen.class);
        Container d = utils.getContainer();
        Intent intent = new Intent(utils.getMultitoolContext(), AppChooser.class);
        screen.startActivityForResult(intent, create, d.getId() + "/" + getClass().getName());
    }

    @Override
    public void onActivityResult(int resultCode, Intent data, String token) {
        if (resultCode == Activity.RESULT_OK) {
            Container d = utils.getActiveScreen().getContainerById(Integer.parseInt(token));
            Script resume = utils.installRegisterScript();
            Script pause = utils.installUnregisterScript();
            Shortcut item = d.addShortcut("0", new Intent(), 0, 0);
            Intent intent = data.getParcelableExtra(Intent.EXTRA_INTENT);
            item.setTag(TAG_PACKAGE, intent.getComponent().getPackageName());
            item.getProperties()
                    .edit()
                    .setBoolean(PropertySet.ITEM_ON_GRID, false)
                    .setBoolean(PropertySet.SHORTCUT_ICON_VISIBILITY, false)
                    .setBoolean(PropertySet.SHORTCUT_LABEL_VISIBILITY, true)
                    .setBoolean(PropertySet.ITEM_ENABLED, false)
                    .setEventHandler(PropertySet.ITEM_RESUMED, EventHandler.RUN_SCRIPT, resume.getId() + "/" + BadgeListener.class.getName())
                    .setEventHandler(PropertySet.ITEM_PAUSED, EventHandler.RUN_SCRIPT, String.valueOf(pause.getId()))
                    .commit();
            utils.centerOnTouch(item);
        }
    }
}

File: app/src/main/java/com/faendir/lightning_launcher/multitool/badge/BadgeSetup.kt
package com.faendir.lightning_launcher.multitool.badge

import android.app.Activity
import android.content.Intent
import androidx.annotation.Keep
import com.faendir.lightning_launcher.multitool.proxy.*

/**
 * @author lukas
 * @since 08.07.18
 */
@Keep
class BadgeSetup(private val utils: Utils) : JavaScript.Setup, JavaScript.ActivityResult {

    override fun setup() {
        val create = utils.installActivityResultScript()
        val screen = ProxyFactory.cast(utils.activeScreen, ActivityScreen::class.java)
        val d = utils.container
        val intent = Intent(utils.multitoolContext, AppChooser::class.java)
        screen.startActivityForResult(intent, create, javaClass.name + "/" + d.id)
    }

    override fun onActivityResult(resultCode: Int, data: Intent, token: String) {
        if (resultCode == Activity.RESULT_OK) {
            val d = utils.activeScreen.getContainerById(Integer.parseInt(token))
            val resume = utils.installRegisterScript()
            val pause = utils.installUnregisterScript()
            val item = d.addShortcut("0", Intent(), 0f, 0f)
            val intent = data.getParcelableExtra<Intent>(Intent.EXTRA_INTENT)

            item.setTag(TAG_PACKAGE, intent.component?.packageName)
            item.properties
                    .edit()
                    .setBoolean(PropertySet.ITEM_ON_GRID, false)
                    .setBoolean(PropertySet.SHORTCUT_ICON_VISIBILITY, false)
                    .setBoolean(PropertySet.SHORTCUT_LABEL_VISIBILITY, true)
                    .setBoolean(PropertySet.ITEM_ENABLED, false)
                    .setEventHandler(PropertySet.ITEM_RESUMED, EventHandler.RUN_SCRIPT, resume.id.toString() + "/" + BadgeListener::class.java.name)
                    .setEventHandler(PropertySet.ITEM_PAUSED, EventHandler.RUN_SCRIPT, pause.id.toString())
                    .commit()
            utils.centerOnTouch(item)
        }
    }

    companion object {
        internal const val TAG_PACKAGE = "package"
    }
}
